llvm/tools/sycl-post-link/ModuleSplitter.cpp

```diff
@@ -424,3 +366,243 @@ std::unique_ptr<ModuleSplitterBase> module_split::getSplitterByMode(
+      }
+      for (auto *UU : U->users())
+        Workqueue.insert(UU);
+    }
+  }
+}
+
+#ifndef _NDEBUG
+
+const char *toString(SyclEsimdSplitStatus S) {
+  switch (S) {
+  case SyclEsimdSplitStatus::ESIMD_ONLY:
+    return "ESIMD_ONLY";
+  case SyclEsimdSplitStatus::SYCL_ONLY:
+    return "SYCL_ONLY";
+  case SyclEsimdSplitStatus::SYCL_AND_ESIMD:
+    return "SYCL_AND_ESIMD";
+  }
+  return "<UNKNOWN_STATUS>";
+}
+
+void tab(int N) {
+  for (int I = 0; I < N; ++I) {
+    llvm::errs() << "  ";
+  }
+}
+
+void dumpEntryPoints(const EntryPointVec &C, const char *msg, int Tab) {
+  tab(Tab);
+  llvm::errs() << "ENTRY POINTS"
+               << " " << msg << " {\n";
+  for (const Function *F : C) {
+    tab(Tab);
+    llvm::errs() << "  " << F->getName() << "\n";
+  }
+  tab(Tab);
+  llvm::errs() << "}\n";
+}
+
+void dumpEntryPoints(const Module &M, bool OnlyKernelsAreEntryPoints,
+                     const char *msg, int Tab) {
+  tab(Tab);
+  llvm::errs() << "ENTRY POINTS (Module)"
+               << " " << msg << " {\n";
+  for (const auto &F : M) {
+    if (isEntryPoint(F, OnlyKernelsAreEntryPoints)) {
+      tab(Tab);
+      llvm::errs() << "  " << F.getName() << "\n";
+    }
+  }
+  tab(Tab);
+  llvm::errs() << "}\n";
+}
+
+void ModuleDesc::renameDuplicatesOf(const Module &MA, StringRef Suff) {
+  Module &MB = getModule();
+#ifndef _NDEBUG
+  DenseSet<StringRef> EntryNamesB;
+  auto It0 = entries().cbegin();
+  auto It1 = entries().cend();
+  std::for_each(It0, It1,
+                [&](const Function *F) { EntryNamesB.insert(F->getName()); });
+#endif // _NDEBUG
+  for (const GlobalObject &GoA : MA.global_objects()) {
+    if (GoA.isDeclaration()) {
+      continue;
+    }
+    StringRef Name = GoA.getName();
+    auto *F = dyn_cast<Function>(&GoA);
+    GlobalObject *GoB =
+        F ? cast_or_null<GlobalObject>(MB.getFunction(Name))
+          : cast_or_null<GlobalObject>(MB.getGlobalVariable(Name));
+
+    if (!GoB || GoB->isDeclaration()) {
+      // function or variable is not shared or is a declaration in MB
+      continue;
+    }
+#ifndef _NDEBUG
+    if (F) {
+      // this is a shared function, must not be an entry point:
+      assert(!EntryNamesB.contains(Name));
+    }
+#endif // _NDEBUG
+    // rename the global object in MB:
+    GoB->setName(Name + Suff);
+  }
+}
+
+void ModuleDesc::assignESIMDProperty() {
+  if (EntryPoints.isEsimd()) {
+    Props.HasEsimd = SyclEsimdSplitStatus::ESIMD_ONLY;
+  } else if (EntryPoints.isSycl()) {
+    Props.HasEsimd = SyclEsimdSplitStatus::SYCL_ONLY;
+  } else {
+    Props.HasEsimd = SyclEsimdSplitStatus::SYCL_AND_ESIMD;
+  }
+#ifndef _NDEBUG
+  verifyESIMDProperty();
+#endif // _NDEBUG
+}
+
+#ifndef _NDEBUG
+void ModuleDesc::verifyESIMDProperty() const {
+  if (Props.HasEsimd == SyclEsimdSplitStatus::SYCL_AND_ESIMD) {
+    return; // nothing to verify
+  }
+  // Verify entry points:
+  for (const auto *F : entries()) {
+    const bool IsESIMDFunction = isESIMDFunction(*F);
```

Comment:
Consider to add "(void)IsESIMDFunction;" or wrap this line under macro like #ifdef DEBUG to avoid unused variable warning.